From ee3b0dc05c6fb26b20b51fd5f55bc2f32a8063e8 Mon Sep 17 00:00:00 2001
From: Lepton Wu <lepton@chromium.org>
Date: Thu, 13 Aug 2020 19:14:13 -0700
Subject: [PATCH] util/ralloc: fix ralloc alignment.

On some malloc implementation, malloc just aligns to size of object.
To make sure ralloc_header always starts at the wanted alignment,
just force the size is aligned to the alignment of ralloc_header.
This fixes crashed on instruction like "movaps %xmm0,0x10(%rax)" which
requires aligned memory access under some malloc implementation.

Signed-off-by: Lepton Wu <lepton@chromium.org>
Change-Id: I1d22bd1814d407e44e3972e65cb980832cedaa89
---
 src/util/ralloc.c | 11 +++++++++--
 1 file changed, 9 insertions(+), 2 deletions(-)

diff --git a/src/util/ralloc.c b/src/util/ralloc.c
index 0d20223d98f..65defe8f5c3 100644
--- a/src/util/ralloc.c
+++ b/src/util/ralloc.c
@@ -28,6 +28,8 @@
 #include <string.h>
 #include <stdint.h>
 
+#include "util/u_math.h"
+
 /* Some versions of MinGW are missing _vscprintf's declaration, although they
  * still provide the symbol in the import library. */
 #ifdef __MINGW32__
@@ -53,10 +55,13 @@ _CRTIMP int _vscprintf(const char *format, va_list argptr);
  */
 struct
 #ifdef _MSC_VER
+ #define RALLOC_HEADER_ALIGNMENT 8
  __declspec(align(8))
 #elif defined(__LP64__)
+ #define RALLOC_HEADER_ALIGNMENT 16
  __attribute__((aligned(16)))
 #else
+ #define RALLOC_HEADER_ALIGNMENT 8
  __attribute__((aligned(8)))
 #endif
    ralloc_header
@@ -116,7 +121,8 @@ ralloc_context(const void *ctx)
 void *
 ralloc_size(const void *ctx, size_t size)
 {
-   void *block = malloc(size + sizeof(ralloc_header));
+   void *block = malloc(align64(size + sizeof(ralloc_header),
+				RALLOC_HEADER_ALIGNMENT));
    ralloc_header *info;
    ralloc_header *parent;
 
@@ -163,7 +169,8 @@ resize(void *ptr, size_t size)
    ralloc_header *child, *old, *info;
 
    old = get_header(ptr);
-   info = realloc(old, size + sizeof(ralloc_header));
+   info = realloc(old, align64(size + sizeof(ralloc_header),
+			       RALLOC_HEADER_ALIGNMENT));
 
    if (info == NULL)
       return NULL;
-- 
2.28.0.220.ged08abb693-goog

